### PR TITLE
Issue #78 Add preflight code for dnsmasq crc.conf

### LINKED ISSUE #78: Add a preflight check for `/etc/NetworkManager/dnsmasq.d/crc.conf` for linux
We are right now not checking if `/etc/NetworkManager/dnsmasq.d/crc.conf` exist and network manager using it, this should be part of preflight checks.

### LINKED ISSUE #94: Add functionality to interact with systemd services
Needed for #92

### PR DESCRIPTION
**Pull request by gbraad**
Fixes #78 
Fixes #94 
Closes #85

### CODE REVIEW

File: pkg/crc/preflight/preflight_checks_linux.go [hunk 1]
```diff
@@ -455,3 +460,31 @@ func fixLibvirtCrcNetworkActive() (bool, error) {
...
+	_, err := os.Stat(crcDnsmasqConfigPath)
+	if err != nil {
+		return false, fmt.Errorf("File not found: %s: %s", crcDnsmasqConfigPath, err.Error())
+	}
+	config, err := ioutil.ReadFile(crcDnsmasqConfigPath)
+	if err != nil {
+		return false, fmt.Errorf("Error opening file: %s: %s", crcDnsmasqConfigPath, err.Error())
+	}
+	if !bytes.Equal(config, c) {
+		return false, fmt.Errorf("Config file contains changes: %s: %s", crcDnsmasqConfigPath)
```
Comment (praveenkumar) [inline pkg/crc/preflight/preflight_checks_linux.go]: You forget to put the variable to second `%s` I think it should be `fmt.Errorf("Config file contains changes: %s: %s", crcDnsmasqConfigPath, string(config))

File: pkg/crc/preflight/preflight_checks_linux.go [hunk 2]
```diff
@@ -455,3 +460,31 @@ func fixLibvirtCrcNetworkActive() (bool, error) {
...
+func fixCrcDnsmasqConfigFile() (bool, error) {
+	cmd := exec.Command("sudo", "tee", crcDnsmasqConfigPath)
+	cmd.Stdin = strings.NewReader(crcDnsmasqConfig)
+	buf := new(bytes.Buffer)
+	cmd.Stderr = buf
+	err := cmd.Run()
+	if err != nil {
+		return false, fmt.Errorf("Failed to write dnsmasq config file: %s: %s: %v", crcDnsmasqConfigPath, buf.String(), err)
+	}
+	return true, err
```
Comment (praveenkumar) [inline pkg/crc/preflight/preflight_checks_linux.go]: @gbraad Problem is just adding or changing the file content not going to work until you reload the NetworkManager service which should be part of this function.

File: pkg/crc/preflight/preflight_checks_linux.go (hunk 1 above)
Comment (gbraad) [reply pkg/crc/preflight/preflight_checks_linux.go]: Hmm, that was missed in the earlier PR? will change...

File: pkg/crc/preflight/preflight_checks_linux.go (hunk 2 above)
Comment (gbraad) [reply pkg/crc/preflight/preflight_checks_linux.go]: will add a shared SystemD control function like we used in minishift

Comment (gbraad) [reply pkg/crc/preflight/preflight_checks_linux.go]: Done

File: pkg/crc/preflight/preflight_checks_linux.go (hunk 1 above)
Comment (gbraad) [reply pkg/crc/preflight/preflight_checks_linux.go]: Removed to dump the contents. Seems very unnecessary.

**Comment by gbraad** [comment]
@praveenkumar updated

File: pkg/crc/preflight/preflight_checks_linux.go [hunk 3]
```diff
@@ -27,10 +28,15 @@ const (
...
 	libvirtDriverDownloadURL = "https://github.com/code-ready/machine-driver-libvirt/releases/download/0.9.1/crc-driver-libvirt"
 	defaultPoolSize          = "20 GB"
+	crcDnsmasqConfigFile     = "crc.conf"
 )
 
 var (
 	libvirtDriverBinaryPath = filepath.Join(driverBinaryDir, libvirtDriverCommand)
+	crcDnsmasqConfigPath    = filepath.Join(string(filepath.Separator), "etc", "NetworkManager", "dnsmasq.d", crcDnsmasqConfigFile)
+	crcDnsmasqConfig        = `server=/tt.testing/192.168.126.1
+address=/apps.tt.testing/192.168.126.11
```
Comment (anjannath) [inline pkg/crc/preflight/preflight_checks_linux.go]: Now that i see this again, will this always be static for linux/libvirt?

Comment (gbraad) [reply pkg/crc/preflight/preflight_checks_linux.go]: For now... usuable as-is.

File: pkg/crc/preflight/preflight_checks_linux.go [hunk 4]
```diff
@@ -272,15 +278,15 @@ func fixMachineDriverLibvirtInstalled() (bool, error) {
 		return false, err
 	}
 
-	stdOut, stdErr, err := RunWithPrivilege("mkdir", "-p", driverBinaryDir)
+	stdOut, stdErr, err := crcos.RunWithPrivilege("mkdir", "-p", driverBinaryDir)
 	if err != nil {
 		return false, fmt.Errorf("%s %v: %s", stdOut, err, stdErr)
 	}
-	stdOut, stdErr, err = RunWithPrivilege("cp", tempFilePath, libvirtDriverBinaryPath)
+	stdOut, stdErr, err = crcos.RunWithPrivilege("cp", tempFilePath, libvirtDriverBinaryPath)
```
Comment (praveenkumar) [inline pkg/crc/preflight/preflight_checks_linux.go]: 👍

File: pkg/crc/preflight/preflight_checks_linux.go [hunk 5]
```diff
@@ -455,3 +461,38 @@ func fixLibvirtCrcNetworkActive() (bool, error) {
...
+	cmd.Stdin = strings.NewReader(crcDnsmasqConfig)
+	buf := new(bytes.Buffer)
+	cmd.Stderr = buf
+	err := cmd.Run()
+	if err != nil {
+		return false, fmt.Errorf("Failed to write dnsmasq config file: %s: %s: %v", crcDnsmasqConfigPath, buf.String(), err)
+	}
+
+	sd := systemd.NewHostSystemdCommander()
+	_, esd := sd.Restart("NetworkManager")
```
Comment (praveenkumar) [inline pkg/crc/preflight/preflight_checks_linux.go]: Even `reload` with work, in this way you don't need to disconnect to internet, most of the time if you are connected with VPN then `restart` the network make it disconnect.

File: pkg/crc/preflight/preflight_checks_linux.go [hunk 6]
```diff
@@ -455,3 +461,38 @@ func fixLibvirtCrcNetworkActive() (bool, error) {
...
+	if esd != nil {
```
Comment (praveenkumar) [inline pkg/crc/preflight/preflight_checks_linux.go]: Can we have `err` only instead `esd`?

File: pkg/crc/systemd/actions/action.go [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+	Disable
+	Status
+	DaemonReload
+)
+
+func (action Action) String() string {
+	actions := [...]string{
+		"start",
+		"stop",
+		"restart",
```
Comment (praveenkumar) [inline pkg/crc/systemd/actions/action.go]: Add a `reload` also for a `service reload` without restart the service.

File: pkg/crc/systemd/systemd.go [hunk 8]
```diff
@@ -0,0 +1,93 @@
...
+	"github.com/code-ready/crc/pkg/crc/systemd/states"
+)
+
+type SystemdCommander interface {
+	Start(name string) (bool, error)
+	Stop(name string) (bool, error)
+	Restart(name string) (bool, error)
+	Status(name string) (states.State, error)
+	Enable(name string) (bool, error)
+	Disable(name string) (bool, error)
```
Comment (praveenkumar) [inline pkg/crc/systemd/systemd.go]: Expose a `Reload` function also

File: pkg/crc/preflight/preflight_checks_linux.go (hunk 6 above)
Comment (gbraad) [reply pkg/crc/preflight/preflight_checks_linux.go]: did this as the err was declared above. changed the structure which prevents the function-scoped use, which I personally thinks makes it less readable, but allows the same name use.

File: pkg/crc/preflight/preflight_checks_linux.go (hunk 5 above)
Comment (gbraad) [reply pkg/crc/preflight/preflight_checks_linux.go]: added `reload`

**Comment by gbraad** [comment]
@praveenkumar updated

File: pkg/crc/preflight/preflight_checks_linux.go [hunk 9]
```diff
@@ -455,3 +461,36 @@ func fixLibvirtCrcNetworkActive() (bool, error) {
...
+	_, err := os.Stat(crcDnsmasqConfigPath)
+	if err != nil {
+		return false, fmt.Errorf("File not found: %s: %s", crcDnsmasqConfigPath, err.Error())
+	}
+	config, err := ioutil.ReadFile(crcDnsmasqConfigPath)
+	if err != nil {
+		return false, fmt.Errorf("Error opening file: %s: %s", crcDnsmasqConfigPath, err.Error())
+	}
+	if !bytes.Equal(config, c) {
+		return false, fmt.Errorf("Config file contains changes: %s", crcDnsmasqConfigPath)
```
Comment (praveenkumar) [inline pkg/crc/preflight/preflight_checks_linux.go]: 👍

File: pkg/crc/preflight/preflight_checks_linux.go [hunk 10]
```diff
@@ -455,3 +461,36 @@ func fixLibvirtCrcNetworkActive() (bool, error) {
...
+	}
+	return true, nil
+}
+
+func fixCrcDnsmasqConfigFile() (bool, error) {
+	cmd := exec.Command("sudo", "tee", crcDnsmasqConfigPath)
```
Comment (praveenkumar) [inline pkg/crc/preflight/preflight_checks_linux.go]: 👍

File: pkg/crc/systemd/actions/action.go [hunk 11]
```diff
@@ -0,0 +1,33 @@
...
+	Disable
+	Status
+	DaemonReload
+)
+
+func (action Action) String() string {
+	actions := [...]string{
+		"start",
+		"stop",
+		"reload",
```
Comment (praveenkumar) [inline pkg/crc/systemd/actions/action.go]: 👍

File: pkg/crc/preflight/preflight_checks_linux.go [hunk 12]
```diff
@@ -455,3 +461,36 @@ func fixLibvirtCrcNetworkActive() (bool, error) {
...
+	cmd.Stdin = strings.NewReader(crcDnsmasqConfig)
+	buf := new(bytes.Buffer)
+	cmd.Stderr = buf
+	if err := cmd.Run(); err != nil {
+		return false, fmt.Errorf("Failed to write dnsmasq config file: %s: %s: %v", crcDnsmasqConfigPath, buf.String(), err)
+	}
+
+	sd := systemd.NewHostSystemdCommander()
+	if _, err := sd.Restart("NetworkManager"); err != nil {
```
Comment (praveenkumar) [inline pkg/crc/preflight/preflight_checks_linux.go]: This is still not fixed we just need to reload the `NetworkManager` service.

Comment (gbraad) [reply pkg/crc/preflight/preflight_checks_linux.go]: huh? Oh chips... I think I know what happened. I didn't save this file from Sublime. It was edited, but uncommitted... pushing asap.

**Pull request merged**